client/src/main/java/io/pravega/client/stream/impl/ControllerResolverFactory.java

```diff
@@ -137,50 +146,154 @@ public void start(Listener listener) {
-                                        new ResolvedServerInfo(
-                                                new InetSocketAddress(node.getEndpoint(), node.getPort())))
-                                .collect(Collectors.toList()))
-                        .build();
+                if (this.enableDiscovery) {
+                    // Make an RPC call to the bootstrapped controller servers to fetch all active controllers.
+                    final ServerResponse controllerServerList =
+                            this.client.getControllerServerList(ServerRequest.getDefaultInstance());
+                    serverInfoGroup = ResolvedServerInfoGroup.builder()
+                            .addAll(controllerServerList.getNodeURIList()
+                                    .stream()
+                                    .map(node ->
+                                            new ResolvedServerInfo(
+                                                    new InetSocketAddress(node.getEndpoint(), node.getPort())))
+                                    .collect(Collectors.toList()))
+                            .build();
+                } else {
+                    // Resolve the bootstrapped server hostnames to get the set of controllers.
+                    final ArrayList<InetSocketAddress> resolvedAddresses = new ArrayList<>();
+                    this.bootstrapServers.forEach(address -> {
+                        final InetSocketAddress socketAddress = new InetSocketAddress(address.getHostString(),
+                                address.getPort());
+                        if (!socketAddress.isUnresolved()) {
+                            resolvedAddresses.add(socketAddress);
+                        }
+                    });
+                    serverInfoGroup = ResolvedServerInfoGroup.builder().addAll(resolvedAddresses.stream()
+                            .map(ResolvedServerInfo::new)
+                            .collect(Collectors.toList()))
+                            .build();
+                }
 
                 // Update gRPC load balancer with the new set of server addresses.
                 log.info("Updating client with controllers: {}", serverInfoGroup);
                 this.resolverUpdater.onUpdate(Collections.singletonList(serverInfoGroup), Attributes.EMPTY);
-            } catch (StatusRuntimeException e) {
-                log.warn("Failed to fetch controller addresses - {}", e);
-                this.resolverUpdater.onError(e.getStatus());
+
+                // We have found at least one controller endpoint. Repeat discovery after the regular schedule.
+                nextScheduleTimeMS = REFRESH_INTERVAL_MS;
+            } catch (Throwable e) {
+                // Catching all exceptions here since this method should never exit without rescheduling the discovery.
+                if (e instanceof StatusRuntimeException) {
+                    this.resolverUpdater.onError(((StatusRuntimeException) e).getStatus());
```

Comment:
My general preference as atomic reference, because it gives a clean interface and clear semantics. Of course volatile and atomic reference are equivalent, as AtomicReference is just a bunch of utility functions and a volatile member. The other option, if you find for what ever reason you need the synchronization block anyway, is to just assign to a local variable inside of the syncronized block and then make the blocking call outside of it. 

In this case any of the above are fine.